File: 5_Average Option/Option_Average_MonteCarlo.py
from math import log, exp, sqrt
import numpy as np
import time
import multiprocessing

# arithmetic average call
# payoff = max(Save,τ − K, 0)

# time_elapsed ------> t
# time_left_to_Maturity ------> T - t
def average_MC(StAve, St, K, time_elapsed, time_left_to_maturity, r, q, sigma, n_prev, n, sims, rep):
    dt = time_left_to_maturity/n
    times = 0
    means = []
    means = np.ndarray(shape = (rep))
    while times < rep:
        optionValues = np.ndarray(shape = (sims))

        for sim in range(sims):
            stockPrices = np.ndarray(shape = (n+1))
            stockPrices[0] = log(St)
            for i in range(n):
                dlnS = np.random.normal(loc = (r-q-sigma**2/2)*dt, scale = sigma*sqrt(dt))
                stockPrices[i+1] = dlnS
            
            stockPrices = np.cumsum(stockPrices)
            stockPrices = np.exp(stockPrices)

            if time_elapsed == 0:
                callValue = max(np.mean(stockPrices) - K, 0) * exp(-r * time_left_to_maturity)
                optionValues[sim] = callValue
            else:
                payoff = (StAve*(n_prev + 1) + sum(stockPrices[1:]))/(n_prev + n + 1) - K
                callValue = max(payoff, 0)
                optionValues[sim] = callValue

        mean = np.mean(optionValues)
        means[times] = mean
        times += 1

    sdOfRep = np.std(means)
    meanOfRep = np.mean(means)
    upperBound = meanOfRep + 2*sdOfRep
    upperBound = round(upperBound, 6)
    lowerBound = meanOfRep - 2*sdOfRep
    lowerBound = round(lowerBound, 6)
    bounds = [lowerBound, upperBound]
    print("============================================================")
    print(f"Average Option : European Call")
    if time_elapsed == 0:
        print(f'[ Save,t = {StAve} | time elapsed = {time_elapsed} ]')
    else:
        print(f'[ Save,t = {StAve} | time elapsed = {time_elapsed} | previous n  = {n_prev} ]')
    print("------------------------------------------------------------")
    print(f"平均 : {round(meanOfRep, 6)}")
    print(f"標準誤 : {round(sdOfRep, 6)}")
    print(f"九十五趴信賴區間 : {bounds}")
    print()

    return meanOfRep




# main
St = 50
StAve = 50
K = 50
r = 0.1
q = 0.05
sigma = 0.8
time_left_to_maturity = 0.25
sims = 10000
rep = 20
n_prev = 100
n = 100

start = time.perf_counter()
if __name__ == '__main__':
    time_elapsed = 0
    p1 = multiprocessing.Process(target = average_MC, args = [StAve, St, K, time_elapsed, time_left_to_maturity, r, q, sigma, n_prev, n, sims, rep])
    time_elapsed = 0.25
    p2 = multiprocessing.Process(target = average_MC, args = [StAve, St, K, time_elapsed, time_left_to_maturity, r, q, sigma, n_prev, n, sims, rep])

    p1.start()
    p2.start()
    p1.join()
    p2.join()

    finish = time.perf_counter()
    print("============================================================")
    print(f'Process finished in {round(finish - start, 2)} second(s).')

File: 5_Average Option/Option_Average_MonteCarlo.cpp
# include <iostream>
# include <cmath>
# include <vector>
# include <random>
# include <chrono>

using namespace std;
// arithmetic average option with fixed strike
// call payoff = max(Save,τ - K, 0)
// put payoff = max(K - Save,τ, 0)

vector<double> average_MC_call(double StAve, double St, double K, double time_elapsed, double time_left_to_maturity, double r, double q, double sigma, int n_prev, int n, int sims, int rep){
    double dt = time_left_to_maturity/n;
    int times = 0;

    default_random_engine generator;
    generator.seed(chrono::system_clock::now().time_since_epoch().count());
    normal_distribution<double> distribution((r - q - 0.5*(pow(sigma, 2)))*dt, sigma*sqrt(dt));

    vector<double> means;
    while(times < rep){
        vector<double> optionValue;

        for(int i = 0; i < sims; i++){
            vector<double> stockPrices;
            stockPrices.push_back(log(St));

            for(int j = 1; j < n+1; j++){
                double dlnS = distribution(generator);
                double sample = stockPrices[j-1] + dlnS;
                stockPrices.push_back(sample);
            }
            for(int k = 0; k < n+1; k++){
                stockPrices[k] = exp(stockPrices[k]);
            }
            // for each loop, single price path is simulated...

            double callValue;
            if(time_elapsed == 0){
                double sum_stockPrice = 0;
                for(int l = 0; l < n+1; l++){
                    sum_stockPrice += stockPrices[l];
                }
                double mean_stockPrice = sum_stockPrice/(n+1);
                callValue = max(mean_stockPrice - K, 0.0)*exp(-r * time_left_to_maturity);
                optionValue.push_back(callValue);
            }
            else{
                double sum_stockPrice = 0;
                for(int l = 1; l < n+1; l++){
                    sum_stockPrice += stockPrices[l];
                }
                double payoff = (StAve*(n_prev + 1) + sum_stockPrice)/(n_prev + n + 1) - K;
                callValue = max(payoff, 0.0) * exp(-r * time_left_to_maturity);
                optionValue.push_back(callValue);
            }
        }
        double sum = 0;
        for(int i = 0; i < sims; i++){
            sum += optionValue[i];
        }
        double mean = sum/sims;
        means.push_back(mean);
        times += 1;
        }
    
    double sum_means = 0;
    for(int l = 0; l < rep; l++){
        sum_means += means[l];
    }
    double meanOfRep = sum_means/rep;
    
    double var = 0.0;
    for(int n = 0; n < rep; n++){
        var += (means[n] - meanOfRep) * (means[n] - meanOfRep);
    }
    var = var/rep;
    double sdOfRep = sqrt(var);
    double upper = meanOfRep + 2*sdOfRep;
    double lower = meanOfRep - 2*sdOfRep;
    vector<double> results = {meanOfRep, lower, upper};
    cout << "==================================================" << endl;
    cout << "Average Option : European call " << endl;
    cout << "[ Save,t = "<< StAve << " | " << "time_elapsed = " << time_elapsed << " ]" << endl;
    cout << "--------------------------------------------------" << endl;
    cout << "mean : " << meanOfRep << endl;
    cout << "standard error : " << sdOfRep << endl;
    cout << "0.95 confidence interval : [ " << lower << ", " << upper << " ]" << endl;
    cout << endl;
    return results;
}

vector<double> average_MC_put(double StAve, double St, double K, double time_elapsed, double time_left_to_maturity, double r, double q, double sigma, int n_prev, int n, int sims, int rep){
    double dt = time_left_to_maturity/n;
    int times = 0;

    default_random_engine generator;
    generator.seed(chrono::system_clock::now().time_since_epoch().count());
    normal_distribution<double> distribution((r - q - 0.5*(pow(sigma, 2)))*dt, sigma*sqrt(dt));

    vector<double> means;
    while(times < rep){
        vector<double> optionValue;

        for(int i = 0; i < sims; i++){
            vector<double> stockPrices;
            stockPrices.push_back(log(St));

            for(int j = 1; j < n+1; j++){
                double dlnS = distribution(generator);
                double sample = stockPrices[j-1] + dlnS;
                stockPrices.push_back(sample);
            }
            for(int k = 0; k < n+1; k++){
                stockPrices[k] = exp(stockPrices[k]);
            }
            // for each loop, single price path is simulated...

            double putValue;
            if(time_elapsed == 0){
                double sum_stockPrice = 0;
                for(int l = 0; l < n+1; l++){
                    sum_stockPrice += stockPrices[l];
                }
                double mean_stockPrice = sum_stockPrice/(n+1);
                putValue = max(K - mean_stockPrice, 0.0)*exp(-r * time_left_to_maturity);
                optionValue.push_back(putValue);
            }
            else{
                double sum_stockPrice = 0;
                for(int l = 1; l < n+1; l++){
                    sum_stockPrice += stockPrices[l];
                }
                double payoff = K - (StAve*(n_prev + 1) + sum_stockPrice)/(n_prev + n + 1);
                putValue = max(payoff, 0.0) * exp(-r * time_left_to_maturity);
                optionValue.push_back(putValue);
            }
        }
        double sum = 0;
        for(int i = 0; i < sims; i++){
            sum += optionValue[i];
        }
        double mean = sum/sims;
        means.push_back(mean);
        times += 1;
        }
    
    double sum_means = 0;
    for(int l = 0; l < rep; l++){
        sum_means += means[l];
    }
    double meanOfRep = sum_means/rep;
    
    double var = 0.0;
    for(int n = 0; n < rep; n++){
        var += (means[n] - meanOfRep) * (means[n] - meanOfRep);
    }
    var = var/rep;
    double sdOfRep = sqrt(var);
    double upper = meanOfRep + 2*sdOfRep;
    double lower = meanOfRep - 2*sdOfRep;
    vector<double> results = {meanOfRep, lower, upper};
    cout << "==================================================" << endl;
    cout << "Average Option : European put " << endl;
    cout << "[ Save,t = "<< StAve << " | " << "time_elapsed = " << time_elapsed << " ]" << endl;
    cout << "--------------------------------------------------" << endl;
    cout << "mean : " << meanOfRep << endl;
    cout << "standard error : " << sdOfRep << endl;
    cout << "0.95 confidence interval : [ " << lower << ", " << upper << " ]" << endl;
    cout << endl;
    return results;
}





int main(){
    double St = 50;
    double StAve = 50;
    double K = 50;
    double r = 0.1;
    double q = 0.05;
    double sigma = 0.8;
    double time_left_to_maturity = 0.25;
    int sims = 10000;
    int rep = 20;
    int n_prev = 100;
    int n = 100;

    average_MC_call(StAve, St, K, 0, time_left_to_maturity, r, q, sigma, n_prev, n, sims, rep);
    average_MC_call(StAve, St, K, 0.25, time_left_to_maturity, r, q, sigma, n_prev, n, sims, rep);

    average_MC_put(StAve, St, K, 0, time_left_to_maturity, r, q, sigma, n_prev, n, sims, rep);
    average_MC_put(StAve, St, K, 0.25, time_left_to_maturity, r, q, sigma, n_prev, n, sims, rep);

    return 0;
}